4480 sns Workflow (#4767)
awsdocs/aws-doc-sdk-examples: cpp/example_code/cross-service/topics_and_queues/README.md

# ***Publish and subscribe to topics using filters and queues***

## Overview

> Publish and subscribe is a mechanism for passing information. It’s used in social media, and it’s also used internally in software applications. A producer publishes a message, and the subscribers receive the message. In software, publish and subscribe notifications makes message passing flexible and robust. The producers of messages are decoupled from the consumers of messages.

Reviewer: ```suggestion
Publish and subscribe is a mechanism for passing information. It’s used in social media, and it’s also used internally in software applications. A producer publishes a message, and the subscribers receive the message. In software, publish and subscribe notifications make message passing flexible and robust. The producers of messages are decoupled from the consumers of messages.
```

Use the sample code in this folder to explore publishing and subscribing to a topic by using filters and queues. This tutorial does not create a complete end-to-end application. Instead, you can use it to play around with a publish and subscribe architecture.

You can create an Amazon SNS topic and subscribe two Amazon SQS queues to the topic. You can enable FIFO (First-In-First-Out) queueing, and you can add filtered subscriptions. Then, you can publish messages to the topic and see the results in the queues.

> You can publish and subscribe using just SNS. But combining SNS with SQS gives you more flexibility in how the messages are consumed.

Reviewer: ```suggestion
You can publish and subscribe using Amazon SNS alone. But combining Amazon SNS with Amazon SQS gives you more flexibility in how the messages are consumed.
```

> SNS is a push service. It pushes to endpoints such as email addresses, mobile application endpoints, or SQS queues. (For a full list of endpoints, see [SNS event destinations](https://docs.aws.amazon.com/sns/latest/dg/sns-event-destinations.html)).

Reviewer: ```suggestion
Amazon SNS is a push service. It pushes to endpoints such as email addresses, mobile application endpoints, or SQS queues. (For a full list of endpoints, see [SNS event destinations](https://docs.aws.amazon.com/sns/latest/dg/sns-event-destinations.html)).
```

> With SQS, messages are received from a queue by polling. With polling, the subscriber receives messages by callling a receive message api. Any code can poll the queue. Also, the messages stay in the queue until you delete them. This gives you more flexibility in how the messages are processed.

Reviewer: ```suggestion
With Amazon SQS, messages are received from a queue by polling. With polling, the subscriber receives messages by calling a receive message API. Any code can poll the queue. Also, the messages stay in the queue until you delete them. This gives you more flexibility in how the messages are processed.
```

The sample code builds a command line application that asks you for input. This is implemented in multiple programming languages, and the interface can vary slightly between languages. The following shows the interface for the C++ implementation.

### Create an SNS topic

```
Would you like to work with FIFO topics? (y/n) 
```

You configure FIFO (First-In-First-Out) topics when you create them. Choosing a FIFO topic enables other options, too. To learn more, see [FIFO topics example use case](https://docs.aws.amazon.com/sns/latest/dg/fifo-example-use-case.html).

```

> Would you like to use content-based deduplication instead of a deduplication ID? (y/n)

Reviewer: ```suggestion
Use content-based deduplication instead of a deduplication ID? (y/n)
```

```

Deduplication is only available for FIFO topics. Deduplication prevents the subscriber from responding more than once to events that are determined to be duplicates. If a message gets published to an SNS FIFO topic and it’s found to be a duplicate within the five-minute deduplication interval, the message is accepted but not delivered. For more information, see [Message deduplication for FIFO topics](https://docs.aws.amazon.com/sns/latest/dg/fifo-message-dedup.html).

Content-based deduplication uses a hash of the content as a deduplication ID. If content-based deduplication is not enabled, you must include a deduplication ID with each message.

```
Enter a name for your SNS topic:
```

Topic names can have 1-256 characters. They can contain uppercase and lowercase ASCII letters, numbers, underscores, and hyphens. If you chose a FIFO topic, the application automatically adds a “.fifo” suffix, which is required for FIFO topics.

> ### Create two SQS queues.

Reviewer: ```suggestion
### Create two SQS queues
```

> Now, configure two SQS queues to subscribe to your topic.

Reviewer 2: Optional, but it's a little unclear at this point why there are 2 queues. Should we describe why there are 2?

Author: Good idea. I added a description.

Author: Now, configure two SQS queues to subscribe to your topic. Using a separate queue for each subscriber can be useful. You can customize how messages are consumed and
how messages are filtered, for instance.

```
Enter a name for an SQS queue.
```

Queue names can have 1-80 characters. They can contain uppercase and lowercase ASCII letters, numbers, underscores, and hyphens. If you chose a FIFO topic, the application automatically adds a “.fifo” suffix, which is required for FIFO queues.

```

> Would you like to filter messages for "<queue name>.fifo"s subscription to 

Reviewer: ```suggestion
Filter messages for "<queue name>.fifo"s subscription to 
```

the topic "<topic name>.fifo"?  (y/n)
```

> If you chose FIFO topics, you can add a filter to the queue’s topic subscription. There are many ways to filter a topic. In this example code, you have the option to filter by a predetermined selection of attributes. For more information about filters, see [Message filtering for FIFO topics](https://docs.aws.amazon.com/sns/latest/dg/fifo-message-filtering.html).

Reviewer 2: I notice here it talks about Filters, but it isn't listed in the code comments or the metadata for what the example does. If the user is specifically looking for information on filters, will they be able to find this example?

```
You can filter messages by one or more of the following "tone" attributes.
1. cheerful
2. funny
3. serious
4. sincere

> Enter a number (or enter zero to not add anything more)

Reviewer: ```suggestion
Enter a number (or enter zero to stop adding more).
```

The application now prompts you to add the second queue. Repeat the previous steps for the second queue.

The following diagram shows the topic and queue options.
![Diagram of the options](images/fifo_topics_diagram.png)

After you create the topic and subscribe both queues, the application lets you publish messages to the topic.

```
Enter a message text to publish.
```

All configurations include a message text.

```
Enter a message group ID for this message.
```

If this is a FIFO topic, then you must include a group ID. The group ID can contain up to 128 alphanumeric characters `(a-z, A-Z, 0-9)` and punctuation `(!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~).`
For more information about group IDs, see [Message grouping for FIFO topics](https://docs.aws.amazon.com/sns/latest/dg/fifo-message-grouping.html).

```
Enter a deduplication ID for this message.
```

If this is a FIFO topic and content-based deduplication is not enabled, then you must enter a deduplication ID. The message deduplication ID can contain up to 128 alphanumeric characters `(a-z, A-Z, 0-9)` and punctuation `(!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~).`

> 

Reviewer 2: The text formatting on the special characters looks a little odd here.

Author: I copied this from docs. I am not sure what is odd about it.

```

> Would you like to add an attribute to this message? (y/n) y

Reviewer: ```suggestion
Add an attribute to this message? (y/n) y
```

```

If you added a filter to one of the subscriptions, you can choose to add a filtering attribute to the message.

```
1. cheerful
2. funny
3. serious
4. sincere
Enter a number for an attribute: 
```

Select a number for an attribute.

```

> Would you like to post another message? (y/n)

Reviewer: ```suggestion
Post another message? (y/n)
```

##  ⚠️ Important

* Running this code might result in charges to your AWS account.
* Running the tests might result in charges to your AWS account.
* We recommend that you grant your code least privilege. At most, grant only the minimum permissions required to perform the task. For more information, see [Grant least privilege](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege).
* This code is not tested in every AWS Region. For more information, see [AWS Regional Services](https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services).

## Run the examples

### Prerequisites

Before using the code examples, first complete the installation and setup steps of [Getting started](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/getting-started.html) in the AWS SDK for
C++ Developer Guide.

###  Instructions

This example uses the CMake build system. For information about the CMake build system, see https://cmake.org/.
Many Integrated Development Environments (IDEs) support CMake. If your preferred IDE supports CMake, follow the IDE's instructions to open this CMake project.

> This project can also be built from a command line interface by using the following commands.

Reviewer: ```suggestion
You can also build this project from a command line interface by using the following commands.
```

```
mkdir build 
cd build
cmake —build ..
```

This builds the executable named “run_topics_and_queues”.

> Running this example requires IAM permissions for both SNS and SQS.

Reviewer: ```suggestion
Running this example requires AWS Identity and Access Management (IAM) permissions for both SNS and SQS.
```
